google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java

```diff
@@ -0,0 +1,347 @@
+import com.google.common.annotations.VisibleForTesting;
+
+import com.intellij.facet.Facet;
+import com.intellij.facet.ui.FacetEditorTab;
+import com.intellij.icons.AllIcons.Ide;
+import com.intellij.openapi.fileChooser.FileChooserDescriptorFactory;
+import com.intellij.openapi.options.ConfigurationException;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.ui.Messages;
+import com.intellij.openapi.ui.TextFieldWithBrowseButton;
+import com.intellij.openapi.util.Comparing;
+import com.intellij.openapi.util.io.FileUtil;
+import com.intellij.openapi.vfs.LocalFileSystem;
+import com.intellij.ui.DocumentAdapter;
+
+import org.jetbrains.annotations.Nls;
+import org.jetbrains.annotations.NotNull;
+
+import java.awt.event.ActionEvent;
+import java.awt.event.ActionListener;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.InvalidPathException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import javax.annotation.Nullable;
+import javax.swing.JButton;
+import javax.swing.JComponent;
+import javax.swing.JLabel;
+import javax.swing.JPanel;
+import javax.swing.event.DocumentEvent;
+
+/**
+ * Panel used to configure the Flexible settings.
+ */
+public class FlexibleFacetEditor extends FacetEditorTab {
+
+  private static final AppEngineProjectService APP_ENGINE_PROJECT_SERVICE =
+      AppEngineProjectService.getInstance();
+
+  private JPanel mainPanel;
+  private TextFieldWithBrowseButton yaml;
+  private TextFieldWithBrowseButton dockerfile;
+  private JButton genYamlButton;
+  private JButton genDockerfileButton;
+  private JLabel noDockerfileLabel;
+  private JLabel errorIcon;
+  private JLabel errorMessage;
+  private AppEngineDeploymentConfiguration deploymentConfiguration;
+  private AppEngineHelper appEngineHelper;
+
+  FlexibleFacetEditor(@Nullable AppEngineDeploymentConfiguration deploymentConfiguration,
+      @Nullable Project project) {
+    this.appEngineHelper = new CloudSdkAppEngineHelper(project);
+    this.deploymentConfiguration = deploymentConfiguration;
+
+    yaml.addBrowseFolderListener(
+        GctBundle.message("appengine.flex.config.browse.app.yaml"),
+        null /* description */,
+        project,
+        FileChooserDescriptorFactory.createSingleFileDescriptor().withFileFilter(
+            virtualFile -> Comparing.equal(virtualFile.getExtension(), "yaml")
+                || Comparing.equal(virtualFile.getExtension(), "yml")
+        )
+    );
+
+    yaml.getTextField().getDocument().addDocumentListener(new DocumentAdapter() {
+      @Override
+      protected void textChanged(DocumentEvent event) {
+        toggleDockerfileSection();
+        validateConfiguration();
+      }
+    });
+
+    dockerfile.getTextField().getDocument().addDocumentListener(new DocumentAdapter() {
+      @Override
+      protected void textChanged(DocumentEvent event) {
```

Comment:
apparently in the UI thread it's OK. http://www.jetbrains.org/intellij/sdk/docs/basics/architectural_overview/general_threading_rules.html

I don't understand the model here well enough to know whether you're in one here.